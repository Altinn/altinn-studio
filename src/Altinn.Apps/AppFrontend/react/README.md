## Running the tests

### Lint checks

1. Navigate to the folder `src/Altinn.Apps/AppFrontend/react`.
2. Execute `npm ci`. This step is only nescessary if you have not already done it, or if you change branches.
3. Execute `npm run install-deps`. This step is only nescessary if you have not already done it, or if you change branches.
4. Execute `npm run lint`.

### Unit tests

1. Navigate to the folder `src/Altinn.Apps/AppFrontend/react`.
2. Execute `npm ci`. This step is only nescessary if you have not already done it, or if you change branches.
3. Execute `npm run install-deps`. This step is only nescessary if you have not already done it, or if you change branches.
4. Execute `npm run test`.
