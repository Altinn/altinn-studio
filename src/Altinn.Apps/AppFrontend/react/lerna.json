{
  "command": {
    "bootstrap": {
      "hoist": true
    }
  },
  "packages": [
    "./*"
  ],
  "version": "independent"
}
