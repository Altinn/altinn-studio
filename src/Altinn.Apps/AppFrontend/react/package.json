{
	"name": "app-frontend",
	"private": true,
	"scripts": {
		"install-deps": "lerna bootstrap --hoist --ci",
		"no-ci-hoist": "lerna bootstrap --hoist --no-ci",
		"install-deps-with-lock": "lerna bootstrap",
		"test": "lerna exec --concurrency 1 -- npm test",
		"install-all": "lerna exec -- npm ci",
		"clean-lock": "lerna exec -- rm ./package-lock.json",
		"clean-node-modules": "rm -rf ./node_modules && lerna exec --concurrency 1 -- rm -rf ./node_modules",
		"update-packages": "lerna exec --concurrency 1 -- npm update --dev",
		"list-outdated": "lerna run list-outdated"
	},
	"devDependencies": {
		"@typescript-eslint/eslint-plugin": "~4.8.2",
		"@typescript-eslint/parser": "~4.8.2",
		"eslint": "~7.14.0",
		"eslint-config-airbnb": "~18.2.1",
		"eslint-plugin-import": "~2.22.1",
		"eslint-plugin-jsx-a11y": "~6.4.1",
		"eslint-plugin-react": "~7.21.5",
		"eslint-plugin-react-hooks": "^4.2.0",
		"ini": "1.3.5",
		"lerna": "3.22.1"
	}
}
