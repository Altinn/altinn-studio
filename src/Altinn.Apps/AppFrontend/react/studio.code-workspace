{
	"folders": [
		{
			"path": "."
		}
  ],
  "settings": {
    "typescript.surveys.enabled": false,
    "typescript.tsdk": "node_modules/typescript/lib",
    "eslint.alwaysShowStatus": true,
    "eslint.validate": [
      "javascript",
      "javascriptreact",
      { "language": "typescript", "autoFix": true },
      { "language": "typescriptreact", "autoFix": true}
    ]
  }
}
