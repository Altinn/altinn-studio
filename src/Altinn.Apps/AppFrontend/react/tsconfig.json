{
  "compilerOptions": {
    "outDir": "compiled",
    "module": "commonjs",
    "target": "es5",
    "lib": [
      "es6",
      "dom",
      "es7",
      "esnext"
    ],
    "sourceMap": true,
    "allowJs": true,
    "jsx": "react",
    "moduleResolution": "node",
    "rootDir": "./",
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noImplicitAny": true,
    "strictNullChecks": false,
    "noUnusedLocals": true
  },
  "include": [
    "**/*.ts",
    "**/*.tsx",
  ],
  "exclude": [
    "node_modules",
    "build",
    "scripts",
    "acceptance-tests",
    "webpack",
    "jest",
    "setupTests.ts",
    "templates",
    "**/setupTests.ts"
  ],
}
