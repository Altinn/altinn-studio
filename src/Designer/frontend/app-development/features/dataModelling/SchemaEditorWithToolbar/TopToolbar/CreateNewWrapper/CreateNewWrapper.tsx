import type { ChangeEvent, KeyboardEvent } from 'react';
import React, { useState } from 'react';
import classes from './CreateNewWrapper.module.css';
import { useTranslation } from 'react-i18next';
import { PlusIcon } from 'libs/studio-icons/src';
import { StudioButton, StudioPopover, StudioTextfield } from '@studio/components-legacy';
import { useValidateSchemaName } from 'app-shared/hooks/useValidateSchemaName';
import { useCreateDataModelMutation } from '../../../../../hooks/mutations';
import type { DataModelMetadata } from 'app-shared/types/DataModelMetadata';
import { extractModelNamesFromMetadataList } from '../../../../../utils/metadataUtils';
import { useAppMetadataQuery } from 'app-shared/hooks/queries';
import { useStudioEnvironmentParams } from 'app-shared/hooks/useStudioEnvironmentParams';
import { extractDataTypeNamesFromAppMetadata } from '../utils/validationUtils';

export interface CreateNewWrapperProps {
  disabled: boolean;
  isCreateNewOpen: boolean;
  createPathOption?: boolean;
  dataModels: DataModelMetadata[];
  setIsCreateNewOpen: (open: boolean) => void;
}

export function CreateNewWrapper({
  disabled,
  createPathOption = false,
  isCreateNewOpen,
  dataModels,
  setIsCreateNewOpen,
}: CreateNewWrapperProps) {
  const { org, app } = useStudioEnvironmentParams();
  const { data: appMetadata } = useAppMetadataQuery(org, app);
  const { mutate: createDataModel } = useCreateDataModelMutation();
  const dataModelNames = extractModelNamesFromMetadataList(dataModels);
  const dataTypeNames = extractDataTypeNamesFromAppMetadata(appMetadata);
  const { validateName, nameError, setNameError } = useValidateSchemaName(
    dataModelNames,
    dataTypeNames,
  );
  const [newModelName, setNewModelName] = useState('');
  const { t } = useTranslation();

  const isConfirmButtonActivated = newModelName && !nameError;
  const relativePath = createPathOption ? '' : undefined;

  const handleNameChange = (e: ChangeEvent<HTMLInputElement>) => {
    const name = e.target.value || '';
    setNewModelName(name);
    validateName(name);
  };

  const handleConfirm = () => {
    createDataModel({
      name: newModelName,
      relativePath,
    });

    handleOpenChange();
  };

  const handleKeyUp = (e: KeyboardEvent) => {
    if (e.key === 'Enter' && isConfirmButtonActivated) {
      handleConfirm();
    }
  };

  const handleOpenChange = () => {
    setIsCreateNewOpen(!isCreateNewOpen);
    setNewModelName('');
    setNameError('');
  };

  return (
    <StudioPopover open={isCreateNewOpen} onClose={handleOpenChange}>
      <StudioPopover.Trigger
        id='create-new-data-model-button'
        disabled={disabled}
        variant='tertiary'
        onClick={handleOpenChange}
        size='small'
      >
        {<PlusIcon />}
        {t('general.create_new')}
      </StudioPopover.Trigger>
      <StudioPopover.Content className={classes.popover}>
        <StudioTextfield
          id='newModelInput'
          label={t('schema_editor.create_model_description')}
          onChange={handleNameChange}
          onKeyUp={handleKeyUp}
          error={nameError}
          autoFocus
        />
        <StudioButton
          color='second'
          onClick={handleConfirm}
          disabled={!isConfirmButtonActivated}
          variant='secondary'
          size='small'
        >
          {t('schema_editor.create_model_confirm_button')}
        </StudioButton>
      </StudioPopover.Content>
    </StudioPopover>
  );
}
