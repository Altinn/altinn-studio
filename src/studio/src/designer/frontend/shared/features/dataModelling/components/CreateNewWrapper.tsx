import React, { useEffect, useState } from 'react';
import { TopToolbarButton } from '@altinn/schema-editor/index';
import {
  Button,
  ButtonColor,
  ButtonVariant,
  ErrorMessage,
  TextField,
} from '@altinn/altinn-design-system';
import AltinnPopoverSimple from '../../../components/molecules/AltinnPopoverSimple';
import { getLanguageFromKey } from '../../../utils/language';

export interface ICreateNewWrapper {
  language: any;
  createAction: ({
    name,
    relativePath,
  }: {
    name: string;
    relativePath: string | undefined;
  }) => void;
  dataModelNames: string[];
  createPathOption?: boolean;
  disabled: boolean;
  openByDefault?: boolean;
}

export function CreateNewWrapper(props: ICreateNewWrapper) {
  const t = (key: string) => getLanguageFromKey(key, props.language);
  const [createButtonAnchor, setCreateButtonAnchor] = useState(null);
  const [newModelName, setNewModelName] = useState('');
  const [nameError, setNameError] = useState('');
  const [confirmedWithReturn, setConfirmedWithReturn] = useState(false);

  useEffect(() => {
    if (props.openByDefault) {
      setCreateButtonAnchor(document.getElementById('create-new-datamodel-button'));
    }
  }, [props.openByDefault]);

  const relativePath = props.createPathOption ? '' : undefined;
  const onCreateClick = (event: any) => {
    setCreateButtonAnchor(event.currentTarget);
  };
  const nameIsValid = () => newModelName.match(/^[a-zA-Z][a-zA-Z0-9_.\-æÆøØåÅ ]*$/);
  const validateName = () => {
    setNameError(!nameIsValid() ? 'Invalid name' : '');
  };
  const onInputBlur = () => {
    if (confirmedWithReturn) {
      setConfirmedWithReturn(false);
      return;
    }
    validateName();
  };
  const onNameChange = (e: any) => {
    const name = e.target.value || '';
    if (nameError) {
      setNameError('');
    }
    setNewModelName(name);
  };
  const onCreateConfirmClick = () => {
    if (nameError || !newModelName || !nameIsValid()) {
      return;
    }
    if (props.dataModelNames.includes(newModelName)) {
      setNameError(`A model with name ${newModelName} already exists.`);
      return;
    }
    props.createAction({
      name: newModelName,
      relativePath,
    });
    setCreateButtonAnchor(null);
    setNewModelName('');
    setNameError('');
  };
  const handleReturnButtonConfirm = () => {
    validateName();
    onCreateConfirmClick();
    setConfirmedWithReturn(true);
  };
  const onKeyUp = (e: React.KeyboardEvent) => {
    if (e.key === 'Enter') {
      handleReturnButtonConfirm();
    }
  };

  const onCancelCreate = () => {
    setCreateButtonAnchor(null);
    setNewModelName('');
    setNameError('');
  };
  return (
    <>
      <TopToolbarButton
        id='create-new-datamodel-button'
        faIcon='fa fa-plus'
        onClick={onCreateClick}
        hideText={false}
        disabled={props.disabled}
      >
        {t('general.create_new')}
      </TopToolbarButton>
      {createButtonAnchor && (
        <AltinnPopoverSimple
          open={!!createButtonAnchor}
          anchorEl={createButtonAnchor}
          handleClose={onCancelCreate}
          anchorOrigin={{
            vertical: 'bottom',
            horizontal: 'left',
          }}
        >
          <label>{t('schema_editor.create_model_description')}</label>
          <TextField
            id='newModelInput'
            placeholder={t('schema_editor.name')}
            isValid={!nameError}
            onChange={onNameChange}
            onBlur={onInputBlur}
            onKeyUp={onKeyUp}
          />
          {nameError && <ErrorMessage>{t(nameError)}</ErrorMessage>}
          <Button
            color={ButtonColor.Secondary}
            onClick={onCreateConfirmClick}
            style={{ marginTop: 22 }}
            variant={ButtonVariant.Outline}
          >
            {t('schema_editor.create_model_confirm_button')}
          </Button>
        </AltinnPopoverSimple>
      )}
    </>
  );
}
CreateNewWrapper.defaultProps = {
  createPathOption: false,
};
