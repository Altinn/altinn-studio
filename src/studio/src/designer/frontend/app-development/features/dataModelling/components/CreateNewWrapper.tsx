import { Grid, Button } from '@material-ui/core';
import { AddCircleOutline } from '@material-ui/icons';
import AltinnInputField from 'app-shared/components/AltinnInputField';
import AltinnPopoverSimple from 'app-shared/components/molecules/AltinnPopoverSimple';
import { getLanguageFromKey } from 'app-shared/utils/language';
import * as React from 'react';

interface ICreateNewWrapper {
  language: any,
  createAction: (modelName: string) => void,
  buttonClass: string,
  dataModelNames: string[],
}
export default function CreateNewWrapper(props: ICreateNewWrapper) {
  const [createButtonAnchor, setCreateButtonAnchor] = React.useState(null);
  const [newModelName, setNewModelName] = React.useState(null);
  const [nameError, setNameError] = React.useState('');

  const onCreateClick = (event: any) => {
    setCreateButtonAnchor(event.currentTarget);
  };
  const onNewModelNameChanged = (e: any) => {
    const name = e.target.value;
    if (!name.match(/^[a-zA-Z][a-zA-Z0-9_.\-æÆøØåÅ ]*$/)) {
      setNameError('Invalid name');
    } else {
      setNameError('');
      setNewModelName(name);
    }
  };
  const onCreateConfirmClick = () => {
    if (newModelName && newModelName.length > 0) {
      if (props.dataModelNames.includes(newModelName.toLowerCase())) {
        setNameError(`A model with name ${newModelName} already exists.`);
        return;
      }
      props.createAction(newModelName);
      setCreateButtonAnchor(null);
      setNewModelName(null);
    }
  };
  const onCancelCreate = () => {
    setCreateButtonAnchor(null);
    setNewModelName(null);
  };
  return (
    <>
      <Grid item>
        <Button
          id='new-button'
          variant='contained'
          className={props.buttonClass}
          startIcon={<AddCircleOutline />}
          onClick={onCreateClick}
        >
          {getLanguageFromKey('general.create_new', props.language)}
        </Button>
      </Grid>
      <AltinnPopoverSimple
        anchorEl={createButtonAnchor}
        handleClose={onCancelCreate}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'left',
        }}
      >
        <AltinnInputField
          id='newModelInput'
          placeholder='Name'
          btnText='Ok'
          error={nameError}
          inputFieldStyling={{ width: '250px' }}
          onChangeFunction={onNewModelNameChanged}
          onBtnClickFunction={onCreateConfirmClick}
        />
      </AltinnPopoverSimple>
    </>
  );
}
