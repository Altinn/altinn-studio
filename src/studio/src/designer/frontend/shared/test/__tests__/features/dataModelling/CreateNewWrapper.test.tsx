import { mount } from 'enzyme';
import 'jest';
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import CreateNewWrapper from '../../../../features/dataModelling/components/CreateNewWrapper';

describe('>>> CreateNewWrapper.tsx', () => {
  const language = { administration: {} };
  let someValue = 'unchangedValue';
  let wrapper: any = null;
  const changeValue = (v: string) => {
    someValue = v;
  };
  beforeEach(() => {
    someValue = 'unchangedValue';
    wrapper = null;
  });
  const mountComponent = (modelNames: string[], onCreate?: (v: string) => void) => mount(<CreateNewWrapper
    language={language}
    dataModelNames={modelNames} createAction={onCreate}
  />);
  it('+++ Should match snapshot with the least amount of params', () => {
    const rendered = renderer.create(
      <CreateNewWrapper
        language={language}
        dataModelNames={[]} createAction={() => { /* intentional */ }}
      />,
    );
    expect(rendered).toMatchSnapshot();
  });
  it('opens the popup when clicking new', () => {
    act(() => {
      wrapper = mountComponent([]);
    });
    const newButton = wrapper.find('button#new-button');
    expect(wrapper.find('input').length).toBe(0);
    newButton.at(0).simulate('click');
    expect(wrapper.find('input').length).toBe(1);
    const inputField = wrapper.find('div#newModelInput').find('input');
    expect(inputField).toBeTruthy();
  });
  const openDialog = () => {
    const newButton = wrapper.find('button#new-button');
    newButton.at(0).simulate('click');
    return wrapper.find('div#newModelInput').find('input');
  };
  it('executes the on change function', () => {
    act(() => {
      wrapper = mountComponent(['some', 'names'], changeValue);
    });
    const inputField = openDialog();
    inputField.simulate('change', { target: { value: 'new-model' } });
    inputField.simulate('blur');
    const okButton = wrapper.find('#newModelInput').find('button');
    okButton.simulate('click');
    wrapper.update();
    expect(someValue).toBe('new-model');
  });
  it('fails to run create name for an existing model', () => {
    act(() => {
      wrapper = mountComponent(['some', 'existing-model', 'names'], changeValue);
    });
    const inputField = openDialog();
    inputField.simulate('change', { target: { value: 'existing-model' } });
    inputField.simulate('blur');
    const okButton = wrapper.find('#newModelInput').find('button');
    okButton.simulate('click');
    wrapper.update();
    expect(someValue).toBe('unchangedValue');
  });
  it('fails to run create when field is empty', () => {
    act(() => {
      wrapper = mountComponent(['some', 'existing-model', 'names'], changeValue);
    });
    const inputField = openDialog();
    inputField.simulate('change', { target: { value: 'dirty-field' } });
    inputField.simulate('blur');
    inputField.simulate('change', { target: { value: '' } });
    inputField.simulate('blur');
    const okButton = wrapper.find('#newModelInput').find('button');
    okButton.simulate('click');
    wrapper.update();
    expect(someValue).toBe('unchangedValue');
  });
});
