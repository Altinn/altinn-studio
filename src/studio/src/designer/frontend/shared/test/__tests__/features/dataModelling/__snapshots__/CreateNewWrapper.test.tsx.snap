// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> CreateNewWrapper.tsx +++ Should match snapshot with the least amount of params 1`] = `
<div
  className="MuiGrid-root MuiGrid-item"
>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained button"
    disabled={false}
    id="new-button"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      <span
        className="MuiButton-startIcon MuiButton-iconSizeMedium"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
          />
        </svg>
      </span>
      general.create_new
    </span>
  </button>
</div>
`;
